privacyidea/lib/tokens/pushtoken.py

```diff
@@ -561,6 +571,47 @@ def create_challenge(self, transactionid=None, options=None):
         self.challenge_janitor()
         return True, message, db_challenge.transaction_id, attributes
 
+    @check_token_locked
+    def authenticate(self, passw, user=None, options=None):
+        """
+        High level interface which covers the check_pin and check_otp
+        This is the method that verifies single shot authentication.
+        The challenge is send to the smartphone app and privacyIDEA
+        waits for the response to arrive.
+
+        :param passw: the password which could be pin+otp value
+        :type passw: string
+        :param user: The authenticating user
+        :type user: User object
+        :param options: dictionary of additional request parameters
+        :type options: dict
+
+        :return: returns tuple of
+                 1. true or false for the pin match,
+                 2. the otpcounter (int) and the
+                 3. reply (dict) that will be added as
+                    additional information in the JSON response
+                    of ``/validate/check``.
+        :rtype: tuple
+        """
+        otp_counter = -1
+        reply = None
+        pin_match = self.check_pin(passw, user=user, options=options)
+        if pin_match:
+            waiting = int(options.get(PUSH_ACTION.WAIT, 20))
+            # Trigger the challenge
+            _t, _m, transaction_id, _attr = self.create_challenge(options=options)
+            # now we need to check and wait for the response to be answered in the challenge table
+            starttime = time.time()
+            while True:
+                # TODO: It is not clear if in a process/threaded environment like apache the challenges are updated in one thread!
+                otp_counter = self.check_challenge_response(options={"transaction_id": transaction_id})
+                if otp_counter >= 0 or (time.time() - starttime) > waiting:
```

Comment:
This may be nitpicking, but: ``time.time()`` is not guaranteed to always increase. So if the system time is reset while we are in this loop, the loop will never terminate.
So we could do something like
```python
elapsed_time = time.time() - starttime
if otp_counter >=0 or elapsed_time > waiting or elapsed_time < 0:
    break
```
If the system time is reset, this will cause the authentication to fail, but at least we don't loop forever.

Reply:
Now I understood. I like it. Will do.

Reply:
done in d370c8f9